config/docs: update and deprecated (#6879)
tendermint/tendermint: docs/nodes/configuration.md

like the file below, however, double check by inspecting the
`config.toml` created with your version of `tendermint` installed:

> ```toml# This is a TOML config file.

Author: ```suggestion
```toml
# This is a TOML config file.
```

proxy-app = "tcp://127.0.0.1:26658"

# A custom human readable name for this node

> moniker = "ape"

Reviewer: lol...

Author: Subtle hints to who we are

proxy-app = "tcp://127.0.0.1:26658"

# A custom human readable name for this node

> moniker = "Fergalicious.local"

Reviewer 2: I wouldn't

# Mode of Node: full | validator | seed
# * validator node
#   - all reactors
#   - with priv_validator_key.json, priv_validator_state.json
# * full node
#   - all reactors
#   - No priv_validator_key.json, priv_validator_state.json
# * seed node
#   - only P2P, PEX Reactor
#   - No priv_validator_key.json, priv_validator_state.json
mode = "validator"

> # If this node is many blocks behind the tip of the chain, FastSync

Reviewer 2: how many is enough?

Author: not sure why this is an option in the config TBH. The node should always fall back to fast sync if behind. @cmwaters do you know why this is a thing?

Reviewer: If you're running blocksync then the termination conditions are:
1) you verify the block of the highest height amongst all your peers
2) a global timeout

If you have this switched to true you will always attempt block sync first before switching to consensus. At the moment, if you fall behind in consensus you will not go back into blocksync but remain in consensus. This is an issue yet to be resolved and requires some sort of reactor orchestration.

Author: is there any case a user would set this to false?

Reviewer 3: @cmwaters Is the global timeout a param somewhere?

Reviewer: > is there any case a user would set this to false?
If they just wanted to catch up via consensus I guess

Reviewer: > @cmwaters Is the global timeout a param somewhere?

Just checked now and it's not so much a global timeout rather termination is when the last response from a peer is greater than 15 seconds (and yes the 15 seconds part is hardcoded)

filter-peers = false

#######################################################
###       Priv Validator Configuration              ###
#######################################################
[priv-validator]

# Path to the JSON file containing the private key to use as a validator in the consensus protocol
key-file = "config/priv_validator_key.json"

# Path to the JSON file containing the last sign state of a validator
state-file = "data/priv_validator_state.json"

# TCP or UNIX socket address for Tendermint to listen on for
# connections from an external PrivValidator process
# when the listenAddr is prefixed with grpc instead of tcp it will use the gRPC Client
laddr = ""

> # Client certificate generated while creating needed files for secure connection.

Reviewer 3: ```suggestion
# Path to the client certificate generated while creating needed files for secure connection.
```

> # If a remote validator address is provided but no certificate, the connection will be insecure

Reviewer 3: Does this mean that the server defaults to not validating TLS connections if the client does not connect with a cert?

Author: Correct.

client-certificate-file = ""

> # Client key generated while creating certificates for secure connection

Reviewer 3: ```suggestion
# Path to the client key generated while creating certificates for secure connection
```

validator-client-key-file = ""

> # Path Root Certificate Authority used to sign both client and server certificates

Reviewer 3: ```suggestion
# Path to the Root Certificate Authority used to sign both client and server certificates
```

#######################################################
[p2p]

> # Enable the new p2p layer.

Reviewer 3: Can we link to the newly added docs on the new p2p layer?

Reviewer: Nope. I can't even find docs on the old p2p layer other than in the spec. But maybe the p2p layer is an implementation detail that should just be in the spec

#######################################################
[p2p]

> # Enable the new p2p layer.

Reviewer: ```suggestion
# Enable the legacy p2p layer.
```

use-legacy = false

# Select the p2p internal queue
queue-type = "priority"

# Address to listen for incoming connections
laddr = "tcp://0.0.0.0:26656"

# Address to advertise to peers for them to dial
# If empty, will use the same port as the laddr,
# and will introspect on the listener or use UPnP
# to figure out the address. ip and port are required
# example: 159.89.10.97:26656
external-address = ""

# Comma separated list of seed nodes to connect to
# We only use these if we can’t connect to peers in the addrbook
# NOTE: not used by the new PEX reactor. Please use BootstrapPeers instead.
# TODO: Remove once p2p refactor is complete
# ref: https:#github.com/tendermint/tendermint/issues/5670
seeds = ""

# Comma separated list of peers to be added to the peer store
# on startup. Either BootstrapPeers or PersistentPeers are
# needed for peer discovery

> bootstrap-peers = ""

Reviewer: I'm not too sure if we want to put this here, but the other day I was thinking that if someone knows a few nodes that are producing snapshots, they should populate this field with their nodeID's so that they connect with them first. This is kind of like a short term service discovery thing

> # Comma separated list of nodes to keep persistent connections to

Reviewer: Can we also add the expected format of each node i.e. nodeID@ip:port + have an example

#######################################################
[p2p]

# Enable the new p2p layer.

> disable-legacy = false

Reviewer 2: this is now "use-legacy" (defaults to false,)

Author: updated, i didn't pull master

# Select the p2p internal queue
queue-type = "priority"

# Address to listen for incoming connections
laddr = "tcp://0.0.0.0:26656"

# Address to advertise to peers for them to dial
# If empty, will use the same port as the laddr,
# and will introspect on the listener or use UPnP
# to figure out the address. ip and port are required
# example: 159.89.10.97:26656
external-address = ""

# Comma separated list of seed nodes to connect to
# We only use these if we can’t connect to peers in the addrbook
# NOTE: not used by the new PEX reactor. Please use BootstrapPeers instead.
# TODO: Remove once p2p refactor is complete

> # ref: https:#github.com/tendermint/tendermint/issues/5670

Reviewer 2: I think seeds/bootstrap peers are probably treated similarly? Bootstrap peers are a better name for it.

flush-throttle-timeout = "100ms"

# Maximum size of a message packet payload, in bytes

> max-packet-msg-payload-size = 1400

Reviewer 2: why?

Author: this was copied from the generated config.toml. Not sure why it increased

Author: https://github.com/tendermint/tendermint/commit/ced66e4eb5498014866f8fda242b62fb5b55009b here is the commit. Seems it was because we arbitrarily picked 1024.

#######################################################
[tx-index]

# The backend database list to back the indexer.
# If list contains null, meaning no indexer service will be used.
#
# The application will set which txs to index. In some cases a node operator will be able
# to decide which txs to index based on configuration set in the application.
#
# Options:
#   1) "null"
#   2) "kv" (default) - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).
# 		- When "kv" is chosen "tx.height" and "tx.hash" will always be indexed.
#   3) "psql" - the indexer services backed by PostgreSQL.
indexer = ["kv"]

# The PostgreSQL connection configuration, the connection format:

> #   postgresql://<user>:<password>@<host>:<port>/<db>?<opts>

Reviewer 2: there's probably more setup that people need to do in order to use this.

Author: we have a section on how to set it up. Ill link to it

#######################################################
[tx-index]

# The backend database list to back the indexer.

> # If list contains null, meaning no indexer service will be used.

Reviewer 3: Pardon my ignorance, does 'null' here mean the string `"null"`?

Author: it is either, a user can leave it blank or put in "null". I had to check this as well, ill add an excerpt about this to not confuse future people.

# If list contains null, meaning no indexer service will be used.
#
# The application will set which txs to index. In some cases a node operator will be able
# to decide which txs to index based on configuration set in the application.
#
# Options:
#   1) "null"
#   2) "kv" (default) - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).

> #   - When "kv" is chosen "tx.height" and "tx.hash" will always be indexed.

Reviewer 3: (ignorance on my part) Is this not true in postgres mode?

Author: you are right, it is true for both scenarios

- `persistent-peers` = is a list of comma separated peers that you will always want to be connected to. If you're already connected to the maximum number of peers, persistent peers will not be added.
- `pex` = turns the peer exchange reactor on or off. Validator node will want the `pex` turned off so it would not begin gossiping to unknown peers on the network. PeX can also be turned off for statically configured networks with fixed network connectivity. For full nodes on open, dynamic networks, it should be turned on.
- `private-peer-ids` = is a comma-separated list of node ids that will _not_ be exposed to other peers (i.e., you will not tell other peers about the ids in this list). This can be filled with a validator's node id.

Recently the Tendermint Team conducted a refactor of the p2p layer. This lead to multiple config paramters being deprecated and/or replaced. 

We will cover the new and deprecated parameters below.
### New Parameters

> There are three new parameters, which are enabled is use-legacy is set to false.

Reviewer: ```suggestion
There are three new parameters, which are enabled if use-legacy is set to false.
```

> - `queue-type` = sets a type of queue to use in the p2p layer. There are three options available fifo, priority and wdrr. The default is priority

Reviewer 3: ```suggestion
- `queue-type` = sets a type of queue to use in the p2p layer. There are three options available `fifo`, `priority` and `wdrr`. The default is priority
```

> - `bootstrap-peers` = is a list of comma seperated peers which will be used to bootstrap the address book. 

Reviewer 3: What is the format of the peers here?

Reviewer 2: I believe `hostname:port`

Reviewer: Yup, same as persistent peers

- `max-connections` = is the max amount of allowed inbound and outbound connections.
### Deprecated Parameters

> Note: For Tendermint 0.35, there are two p2p implementations. The old version is used by deafult with the deprecated fields. The new implementation uses different config parameters, explained above.

- `max-num-inbound-peers` = is the maximum number of peers you will accept inbound connections from at one time (where they dial your address and initiate the connection). *This was replaced by `max-connections`*
- `max-num-outbound-peers` = is the maximum number of peers you will initiate outbound connects to at one time (where you dial their address and initiate the connection).*This was replaced by `max-connections`*

> - `unconditional-peer-ids` = is similar to `persistent-peers` except that these peers will be connected to even if you are already connected to the maximum number of peers. This can be a validator node ID on your sentry node. *Depreacted*

Reviewer: ```suggestion
- `unconditional-peer-ids` = is similar to `persistent-peers` except that these peers will be connected to even if you are already connected to the maximum number of peers. This can be a validator node ID on your sentry node. *Deprecated*
```

> - `seeds` = is a list of comma separated seed nodes that you will connect upon a start and ask for peers. A seed node is a node that does not participate in consensus but only helps propagate peers to nodes in the networks *Depreacted*

Reviewer: ```suggestion
- `seeds` = is a list of comma separated seed nodes that you will connect upon a start and ask for peers. A seed node is a node that does not participate in consensus but only helps propagate peers to nodes in the networks *Deprecated, use bootstrap peers instead if starting a fresh Tendermint node*
```

Author: ```suggestion
- `seeds` = is a list of comma separated seed nodes that you will connect upon a start and ask for peers. A seed node is a node that does not participate in consensus but only helps propagate peers to nodes in the networks *Deprecated, replaced by bootstrap peers*
```

## Indexing Settings

Operators can configure indexing via the `[tx_index]` section. The `indexer`
field takes a series of supported indexers. If `null` is included, indexing will
be turned off regardless of other values provided.

### Supported Indexers

#### KV

The `kv` indexer type is an embedded key-value store supported by the main

> underling Tendermint database. Using the `kv` indexer type allows you to query

Reviewer 3: ```suggestion
underlying Tendermint database. Using the `kv` indexer type allows you to query
```
